### PR TITLE
RFC: Contextual Difficulty

### PR DESCRIPTION
**Pull request by teor2345**
[Rendered](https://github.com/teor2345/zebra/blob/contextual-difficulty-rfc/book/src/dev/rfcs/0006-contextual-difficulty.md)

## Status

First draft of reference-level design, focusing on the implementation.
Revised draft of summary, motivation, and guide-level explanation.

I haven't started the test plan yet.
There are some notes in the remaining sections.

## Motivation

We need a design for contextual validation of the proof of work difficulty threshold.

## Solution

The RFC should focus on Contextual Difficulty Validation, and summarise the Difficulty Validation Service Request in enough detail to justify the relevant design decisions.

### Contextual Difficulty Validation - First Alpha

A synchronous contextual verifier that takes a **block header** (not a full block), and returns a Result that says if the difficulty threshold in the header matches the difficulty adjustment specified by the consensus rules.

#### Definitions

- [x] define proof of work, difficulty threshold, work, and cumulative work, and use them consistently
  - [x] update the state RFC with the work definition
- [x] define and specify the relevant chain for the difficulty consensus rules

#### Consensus Rules

- [x] calculate the median proof of work of recent blocks
- [x] calculate the median time interval between recent blocks
- [x] scale the median proof of work based on the median time interval
- [x] allow minimum difficulty blocks on testnet, if the block time gap is larger than the minimum difficulty threshold

#### Guide-Level Explanation

- [x] Revised Draft

#### Detailed Reference Design

- [x] Implementation

#### Module Structure

- [x] implement adjustment calculations in `zebra_chain::work::difficulty` (existing module)
- [x] implement header validation in `zebra_consensus::block::check`
- [x] write module summary

## Follow Up

Difficulty Validation Service Request - see #1166

## Related Work

* RFC issue #1036
* proof of work tracking issue #802 
* state design RFC #902

### CODE REVIEW

**Comment by teor2345** [comment]
@yaahc can you do a quick review to make sure that the definitions and guide-level explanation fit with the current state service design?

In particular, the "relevant chain" for contextual validation needs to work with the non-finalised and finalised state.

I'll also get others to have a look when they have some time.

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+* **median block span**: The 11 most recent blocks from a chosen tip, typically
+  the relevant tip.
+
+* **target spacing**: 150 seconds per block before Blossom activation, 75 seconds
+  per block from Blossom activation onwards.
+
+* **mean target difficulty**: The arithmetic mean of the difficulty thresholds
+  of the blocks in the averaging window.
+
+**median timespan**: The average number of seconds taken to generate the blocks
```
Comment (yaahc) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
* **median timespan**: The average number of seconds taken to generate the blocks
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+  in the averaging window. Calculated using the difference of median block spans
+  in and after the averaging window, then damped and bounded.
+
+**target timespan**: The target spacing for an averaging window's worth of
```
Comment (yaahc) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
* **target timespan**: The target spacing for an averaging window's worth of
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+  chain is usually 100 blocks (the reorg limit) above the finalized tip. But it can
+  be lower during the initial sync, and after a chain reorganization, if the new
+  best chain is at a lower height.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The difficulty threshold for the next block is calculated using the difficulty
+thresholds and times of recent blocks. Zcash uses the most recent 28 blocks in
+the **relevant chain** in its difficulty adjustment calculations.
```
Comment (yaahc) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: If we ever do that `Network` trait change and get it so we can create customized testnets for individual tests so we can test post sapling blocks without cached state this assumption won't necessarily hold

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+## The relevant chain
+[relevant-chain]: #relevant-chain
+
+The relevant chain can be retrieved from the state service [RFC5] as follows:
+* if the previous block is the finalized tip:
+  * get recent blocks from the finalized state
+* if the previous block is in the non-finalized state:
+  * get recent blocks from the relevant chain, then
+  * get recent blocks from the finalized state, if required
```
Comment (yaahc) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: If the difficiulty threshold calculation is being done outside the state service we shouldn't talk about finalized vs non-finalized state. We can just say "get the previous blocks from the state" and let the state handle abstracting the finalized and non-finalized state into a single continuous state for outside observers.

If this is done during contextual validation (seems like it should be) in `CommitBlock` then nvm.

Comment (teor2345) [reply book/src/dev/rfcs/0006-contextual-difficulty.md]: Good point.

I'll update the RFC so that the changes to the state service are separate from the new contextual validation functions.

File: book/src/dev/rfcs/0006-contextual-difficulty.md (hunk 3 above)
Comment (teor2345) [reply book/src/dev/rfcs/0006-contextual-difficulty.md]: I think you're referring to the assumption about chain length a few lines below this line:
> Since contextual validation is only used for post-Sapling blocks, we can assume
that there will be at least 28 blocks in any relevant chain.

That assumption only holds on mainnet and testnet (and I'll update the RFC to say that).

But since that's the case, we can make that assumption part of the contract of the validation function, and panic if it doesn't hold. Then any tests will have to respect that assumption, and call `CommitFinalizedBlock` on the first 28 blocks.

The other alternative is writing code and tests for an edge case that will never happen on a deployed node. Which feels like wasted effort.

File: book/src/dev/rfcs/0006-contextual-difficulty.md (hunk 4 above)
Comment (teor2345) [reply book/src/dev/rfcs/0006-contextual-difficulty.md]: The validation is done during `CommitBlock` in this RFC.

There might be some future work where we pre-download and pre-validate headers, before downloading the entire block. That would require an extra state service request, but the request would just call the existing state service function used for contextual validation.

I'll sketch that design in the future work section.

File: book/src/dev/rfcs/0006-contextual-difficulty.md (hunk 3 above)
Comment (yaahc) [reply book/src/dev/rfcs/0006-contextual-difficulty.md]: a little unfortunate but that sounds reasonable.

Comment (teor2345) [reply book/src/dev/rfcs/0006-contextual-difficulty.md]: Fixed in 8983be3b9a46272fee088690362663814bf338b6.

File: book/src/dev/rfcs/0006-contextual-difficulty.md (hunk 4 above)
Comment (teor2345) [reply book/src/dev/rfcs/0006-contextual-difficulty.md]: Fixed in 02dd6bc4b7fe2ee243ffbf8d2ee774a568d07e08, turns out there are actually 2 separate contextual checks: the difficulty threshold, and the difficulty filter.

(We are checking the difficulty filter in zebra-consensus, but that's wrong, because testnet minimum difficulty blocks make it a contextual check. I'll fix that in a new PR.)

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 5]
```diff
@@ -119,7 +120,13 @@ TODO:
   - check how zcashd implements signed median time differences
   - open a ticket to update the Zcash spec
 
-## The relevant chain
+## State service interface changes
+[state-service-interface]: #state-service-interface
+
+Contextual validation accesses recent blocks. So we modify the the internal state
+service interface to provide an abstraction for accessing recent blocks.
```
Comment (yaahc) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: I was gonna say it already has this abstraction but then I realized that it doesn't, because the current one only works with best chains

https://github.com/teor2345/zebra/blob/contextual-difficulty-rfc/zebra-state/src/service.rs#L230-L236

Comment (yaahc) [reply book/src/dev/rfcs/0006-contextual-difficulty.md]: Thinking about it I think we might want to add an iterator API to the two state types, then we can write the following

```rust
let blocks = self.mem.chain_iter(&block.header.previous_block_hash).chain(self.sled.chain_iter()).take(28);
// ...
```

Comment (teor2345) [reply book/src/dev/rfcs/0006-contextual-difficulty.md]: Resolved by the implementation in PR #1271

**Comment by teor2345** [comment]
This design is now complete enough for an additional review.

The guide-level and reference-level explanations contain a substantially complete first draft.

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 6]
```diff
@@ -0,0 +1,813 @@
...
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The difficulty threshold for the next block is calculated using the difficulty
+thresholds and times of recent blocks. Zcash uses the most recent 28 blocks in
+the **relevant chain** in its difficulty adjustment calculations.
+
+The difficulty adjustment calculations adjust the **mean target difficulty**,
+based on the difference between the **median timespan** and the
+**target timespan**.
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: TODO:
- ~cut detail from the guide-level explanation - turn it into a summary~
- [x] add a short note explaining how the difficulty filter and difficulty adjustment interact

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 7]
```diff
@@ -0,0 +1,813 @@
...
+Contextual validation is implemented in
+`StateService::check_contextual_validity`, which calls a separate function for
+each contextual validity check.
+
+In Zebra, contextual validation starts after Sapling activation, so we can assume
+that the relevant chain contains at least 28 blocks on Mainnet and Testnet. (And
+panic if this assumption does not hold at runtime.)
+
+For debugging purposes, the candidate block's height, hash, and network should be
+included in a span that is active for the entire contextual validation function.
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Move to the tracking issue, delete from the spec

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 8]
```diff
@@ -0,0 +1,813 @@
...
+value, which uses a custom 32-bit floating-point encoding. Zebra calls this type
+`CompactDifficulty`.
+
+In Zcash, difficulty threshold calculations are performed using unsigned 256-bit
+integers. Rust has no standard `u256` type, but there are a number of crates
+available which implement the required operations on 256-bit integers. Zebra
+abstracts over the alternative `u256` implementations using its
+`ExpandedDifficulty` type.
+
+In Zcash, time values are 32-bit integers. But the difficulty adjustment
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
In Zcash, time values are unsigned 32-bit integers. But the difficulty adjustment
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 9]
```diff
@@ -0,0 +1,813 @@
...
+equivalent results to `zcashd` and the Zcash specification.
+
+In Zcash block headers, difficulty thresholds are stored as a "compact" `nBits`
+value, which uses a custom 32-bit floating-point encoding. Zebra calls this type
+`CompactDifficulty`.
+
+In Zcash, difficulty threshold calculations are performed using unsigned 256-bit
+integers. Rust has no standard `u256` type, but there are a number of crates
+available which implement the required operations on 256-bit integers. Zebra
+abstracts over the alternative `u256` implementations using its
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
abstracts over the chosen `u256` implementation using its
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 10]
```diff
@@ -0,0 +1,813 @@
...
+calculations include time subtractions which could overflow an unsigned type, so
+they are performed using signed 64-bit integers in `zcashd`.
+
+Zebra parses the `header.time` field into a `DateTime<Utc>`. Conveniently, the
+`chrono::DateTime<_>::timestamp()` function returns `i64` values. So Zebra can do
+its signed time calculations using `i64` values internally.
+
+Note: `i32` is an unsuitable type for signed time calculations. It is
+theoretically possible for the time gap between blocks to be larger than
+`2^31 - 1`, because those times are provided by miners. Even if the median time
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
`i32::MAX`, because those times are provided by miners. Even if the median time
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 11]
```diff
@@ -0,0 +1,813 @@
...
+impl ExactSizeIterator for Iter<'_> { ... }
+impl FusedIterator for Iter<'_> {}
+```
+
+`chain` and `Iter` are implemented in the existing `zebra_state::service` module.
+
+For further details, see [PR 1271].
+
+[PR 1271]: https://github.com/ZcashFoundation/zebra/pull/1271
+
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Move `StateService::chain` and `Iter` to state RFC

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 12]
```diff
@@ -0,0 +1,813 @@
...
+  * get recent blocks from the finalized state
+* if the previous block is in the non-finalized state:
+  * get recent blocks from the relevant chain, then
+  * get recent blocks from the finalized state, if required
+
+The relevant chain can start at any non-finalized block, or at the finalized tip.
+If the next block is valid, it becomes the new tip of the relevant chain.
+
+In particular, if the previous block is not a chain tip, the relevant chain
+becomes a new chain fork.
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Delete this

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 13]
```diff
@@ -0,0 +1,813 @@
...
+[contextual-validation-design]: #contextual-validation-design
+
+Contextual validation is performed synchronously by the state service, as soon
+as the state has:
+* received the semantically valid next block (via `CommitBlock`), and
+* committed the previous block.
+
+Contextual difficulty validation consists of the following check:
+* the difficulty threshold in the block matches the adjusted difficulty from
+  previous blocks.
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
Contextual difficulty validation consists of the following check:
* the difficulty threshold in the block matches the difficulty adjustment calculated
  from previous blocks.
```

Or maybe:
```
The difficulty adjustment check calculates the correct difficulty threshold
value for a candidate block, and ensures that the block's
`difficulty_threshold` field is equal to that value.
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 14]
```diff
@@ -0,0 +1,813 @@
...
+/// See `new_from_block` for details.
+fn new_from_fields(candidate_time: DateTime<Utc>,
+                   candidate_height: block::Height,
+                   network: Network,
+                   context: &[(CompactDifficulty, DateTime<Utc>); 28])
+                   -> DifficultyAdjustment { ... }
+```
+
+`DifficultyAdjustment` is located in a new
+`zebra_chain::work::difficulty::adjustment` module.
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Make this change for all `DifficultyAdjustment` methods
```suggestion
`zebra_state::service::check::difficulty` module.
```

Actually, let's just move all the modules info to the modules section.

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 15]
```diff
@@ -0,0 +1,813 @@
...
+/// Initialise and return a new `DifficultyAdjustment` using its fields.
+///
+/// Designed for use in tests.
+///
+/// See `new_from_block` for details.
+fn new_from_fields(candidate_time: DateTime<Utc>,
+                   candidate_height: block::Height,
+                   network: Network,
+                   context: &[(CompactDifficulty, DateTime<Utc>); 28])
+                   -> DifficultyAdjustment { ... }
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Delete this

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 16]
```diff
@@ -0,0 +1,813 @@
...
+```rust
+/// Validate the `difficulty_threshold` from a candidate block's header, based
+/// on a `difficulty_adjustment` for that block.
+pub fn difficulty_threshold_is_valid(difficulty_threshold: CompactDifficulty,
+                                     difficulty_adjustment: DifficultyAdjustment,
+                                     -> Result<(), BlockError> { ... }
+```
+
+`difficulty_threshold_is_valid` is located in the existing
+`zebra_consensus::block::check` module.
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
`difficulty_threshold_is_valid` is located in a new `zebra_state::service::check`
module.
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 17]
```diff
@@ -0,0 +1,813 @@
...
+`candidate_difficulty_threshold` is equal to the `difficulty_threshold` value
+calculated using `DifficultyAdjustment::adjusted_difficulty_threshold`.
+
+We implement this function:
+```rust
+/// Validate the `difficulty_threshold` from a candidate block's header, based
+/// on a `difficulty_adjustment` for that block.
+pub fn difficulty_threshold_is_valid(difficulty_threshold: CompactDifficulty,
+                                     difficulty_adjustment: DifficultyAdjustment,
+                                     -> Result<(), BlockError> { ... }
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
/// Validate the `difficulty_threshold` from a candidate block's header, based
/// on a `expected_difficulty` for that block, which is used to calculate
/// `ToCompact(Threshold())`.
pub fn difficulty_threshold_is_valid(difficulty_threshold: CompactDifficulty,
                                     expected_difficulty: DifficultyAdjustment,
                                     -> Result<(), BlockError> { ... }
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 18]
```diff
@@ -0,0 +1,813 @@
...
+chain.
+
+Since the `PoWLimit`s are `2^251 − 1` for Testnet, and `2^243 − 1` for mainnet,
+the sum of these difficulty thresholds will be less than or equal to
+`(2^251 − 1)*17 = 2^255 + 2^251 - 17`. Therefore, this calculation can not
+overflow a `u256` value.
+
+In Zebra, contextual validation starts after Sapling activation, so we can assume
+that the relevant chain contains at least 17 blocks. Therefore, the `PoWLimit`
+case of `MeanTarget()` in the Zcash specification is unreachable.
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Turn into footnote or sub-section?

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 19]
```diff
@@ -0,0 +1,813 @@
...
+that the relevant chain contains at least 28 blocks. Therefore:
+* `max(0, height − PoWMedianBlockSpan)` in the `MedianTime()` calculation
+   simplifies to `height − PoWMedianBlockSpan`,
+* there is always an odd number of blocks in `MedianTime()`, so the median is
+  always the exact middle of the sequence,
+* we only need the candidate block's network upgrade to determine the
+  `AveragingWindowTimespan`, so we don't need the block's network, and
+* we don't need to know the block's height, because all the other uses of
+  `height` in the Zcash specification are implicitly handled by indexing into
+  the `timespan_times` slice.
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Turn into footnote or sub-section?

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 20]
```diff
@@ -0,0 +1,813 @@
...
+and times from the previous 28 blocks in the relevant chain.
+
+These functions also use the candidate block's `height` and `network`.
+
+To make these functions more ergonomic, we create a `DifficultyAdjustment`
+type, and implement the difficulty adjustment calculations as methods on that
+type.
+
+```rust
+struct DifficultyAdjustment {
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Consider renaming to `DifficultyAdjustmentContext`

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 21]
```diff
@@ -0,0 +1,813 @@
...
+#### Test network minimum difficulty implementation
+[test-net-min-difficulty-implementation]: #test-net-min-difficulty-implementation
+
+The testnet minimum difficulty calculation uses the existing
+`NetworkUpgrade::minimum_difficulty_spacing_for_height` function to calculate the
+minimum difficulty gap.
+
+In Zcash, the testnet minimum difficulty rule starts at block 299188, and in
+Zebra, contextual validation starts after Sapling activation. So we can assume
+that there is always a previous block.
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Turn into footnote or sub-section?

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 22]
```diff
@@ -0,0 +1,813 @@
...
+In Zebra, contextual validation starts after Sapling activation, so the genesis
+case of `Threshold()` in the Zcash specification is unreachable.
+
+Note that `zcashd` truncates the `MeanTarget` after the mean calculation, and
+after dividing by `AveragingWindowTimespan`. But as long as there is no overflow,
+this is [equivalent to the single truncation of the final result] in the Zcash
+specification. However, Zebra should follow the order of operations in `zcashd`,
+and use repeated divisions, because that can't overflow.
+
+[equivalent to the single truncation of the final result]: https://math.stackexchange.com/questions/147771/rewriting-repeated-integer-division-with-multiplication
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Turn into footnote or sub-section?

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 23]
```diff
@@ -0,0 +1,813 @@
...
+///
+/// Implements `MedianTime` from the Zcash specification.
+fn median_time(median_block_span_times: &[DateTime<Utc>; 11])
+               -> DateTime<Utc> { ... }
+```
+
+`median_timespan_bounded` and `median_time` are located in a new
+`zebra_chain::work::difficulty::adjustment` module.
+
+Zebra implements the `AveragingWindowTimespan` using the foollowing methods on
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
Zebra implements the `AveragingWindowTimespan` using the following methods on
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 24]
```diff
@@ -0,0 +1,813 @@
...
+* `floor((2^251 − 1) / 2250) * 3366`, which equals
+* `(2^251 − 1) * 132/100`,
+* which is less than `2^252`.
+
+Therefore, this calculation can not overflow a `u256` value. (And even if it did
+overflow, it would be constrained to a valid value by the `PoWLimit` minimum.)
+
+Note that the multiplication by `ActualTimespanBounded` must happen after the
+division by `AveragingWindowTimespan`. Performing the multiplication first
+could overflow.
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Move subsection below implementation details?

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 25]
```diff
@@ -0,0 +1,813 @@
...
+* if the time gap is greater than the testnet minimum difficulty gap:
+  the difficulty threshold is the testnet minimum difficulty threshold.
+
+See [ZIP-208] for details.
+
+Note: There were several errors in the specification of testnet minimum
+difficulty adjustment in ZIPs 205 and 208. The time gap, minimum difficulty
+threshold value, the modification of the `difficulty_threshold` (`nBits`) field,
+and its use in future difficulty adjustments were all incorrect. These errors are
+fixed in [ZIP PR 417] and [ZIP commit 806076c].
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Copy phrasing in ZIPs 205 and 208 - tone this down a bit

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 26]
```diff
@@ -0,0 +1,813 @@
...
+We implement these methods on `DifficultyAdjustment`:
+```rust
+/// Calculate the `difficulty_threshold` for a candidate block, based on the
+/// `candidate_time`, `candidate_height`, `network`, and the
+/// `difficulty_threshold`s and `time`s from the previous
+/// `PoWAveragingWindow + PoWMedianBlockSpan` (28) blocks in the relevant chain.
+///
+/// Implements `ThresholdBits` from the Zcash specification, and the testnet
+/// minimum difficulty adjustment from ZIPs 205 and 208.
+pub fn adjusted_difficulty_threshold(&self) -> CompactDifficulty { ... }
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
pub fn expected_difficulty_threshold(&self) -> CompactDifficulty { ... }
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 27]
```diff
@@ -0,0 +1,813 @@
...
+Note that the multiplication by `ActualTimespanBounded` must happen after the
+division by `AveragingWindowTimespan`. Performing the multiplication first
+could overflow.
+
+#### Block difficulty threshold implementation
+[block-difficulty-threshold-implementation]: #block-difficulty-threshold-implementation
+
+We implement these methods on `DifficultyAdjustment`:
+```rust
+/// Calculate the `difficulty_threshold` for a candidate block, based on the
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
/// Calculate the expected `difficulty_threshold` for a candidate block, based on the
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 28]
```diff
@@ -0,0 +1,813 @@
...
+/// `candidate_height`, `network`, and the relevant `difficulty_threshold`s and
+/// `time`s.
+///
+/// See `adjusted_difficulty_threshold` for details.
+///
+/// Implements `ThresholdBits` from the Zcash specification. (Which excludes the
+/// testnet minimum difficulty adjustment.)
+fn threshold_bits(network: Network,
+                  height: block::Height,
+                  context: &[(CompactDifficulty, DateTime<Utc>); 28])
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
                  relevant_difficulty_thresholds: &[CompactDifficulty; 28],
                  relevant_times: &[DateTime<Utc>; 28],
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 29]
```diff
@@ -0,0 +1,813 @@
...
+[no-action]: #no-action
+
+Zebra could accept invalid, low-difficulty blocks from arbitrary miners. That
+would be a security issue.
+
+# Prior art
+[prior-art]: #prior-art
+
+Discuss prior art, both the good and the bad, in relation to this proposal.
+
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 30]
```diff
@@ -0,0 +1,813 @@
...
+## Alternate consensus parameters
+[alternate-consensus-parameters]: #alternate-consensus-parameters
+
+Any alternate consensus parameters or `regtest` mode would have to respect the constraints set by this design.
+
+In particular:
+  * the `PoWLimit` must be less than `(2^256 - 1) / PoWAveragingWindow` to avoid overflow,
+  * the `PoWAveragingWindow` and `PoWMedianBlockSpan` are fixed by function argument types
+    (at least until Rust gets generic slice lengths), and
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
    (at least until Rust gets const generics), and
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 31]
```diff
@@ -0,0 +1,813 @@
...
+
+Why should we *not* do this?
+
+## Alternate consensus parameters
+[alternate-consensus-parameters]: #alternate-consensus-parameters
+
+Any alternate consensus parameters or `regtest` mode would have to respect the constraints set by this design.
+
+In particular:
+  * the `PoWLimit` must be less than `(2^256 - 1) / PoWAveragingWindow` to avoid overflow,
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]:
```suggestion
  * the `PoWLimit` must be less than or equal to `(2^256 - 1) / PoWAveragingWindow ~= 2^251` to avoid overflow,
```

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 32]
```diff
@@ -0,0 +1,813 @@
...
+  * the `PoWAveragingWindow` and `PoWMedianBlockSpan` are fixed by function argument types
+    (at least until Rust gets generic slice lengths), and
+  * the design eliminates a significant number of edge cases by assuming that difficulty adjustments aren't
+    validated for the first `PoWAveragingWindow + PoWMedianBlockSpan` (28) blocks in the chain.
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md:711]: TODO:
- [x] Add a note about these constraints in the relevant ticket #1135

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 33]
```diff
@@ -0,0 +1,813 @@
...
+This design re-uses existing Zebra code, and follows typical Zebra and Rust design patterns.
+
+## Is this design a good basis for later designs or implementations?
+[good-basis]: #good-basis
+
+The design enables access to recent blocks for contextual validation using a generic
+iterator function.
+
+The design includes specific methods for header-only validation.
+
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Delete this

File: book/src/dev/rfcs/0006-contextual-difficulty.md [hunk 34]
```diff
@@ -0,0 +1,813 @@
...
+## Zcash specification
+[zcash-spec]: #zcash-spec
+
+See the reference-level explanation for prior art and deliberate divergences.
+
+## Bitcoin
+[bitcoin]: #bitcoin
+
+- [ ] TODO: Bitcoin
```
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: Turn into bullets

File: book/src/dev/rfcs/0006-contextual-difficulty.md (hunk 22 above)
Comment (teor2345) [inline book/src/dev/rfcs/0006-contextual-difficulty.md]: See also this clarification comment in `zcashd`:
https://github.com/zcash/zcash/pull/4860/files

File: book/src/dev/rfcs/0006-contextual-difficulty.md (hunk 6 above)
Comment (teor2345) [reply book/src/dev/rfcs/0006-contextual-difficulty.md]: I moved "cut detail from the guide-level explanation - turn it into a summary" into the tracking issue, because it's more work than I expected.

File: book/src/dev/rfcs/0006-contextual-difficulty.md (hunk 20 above)
Comment (teor2345) [reply book/src/dev/rfcs/0006-contextual-difficulty.md]: Renamed to `AdjustedDifficulty`

**Review by yaahc — DISMISSED** [review]
:shipit:

**Review by yaahc — DISMISSED** [review]

**Review by yaahc — DISMISSED** [review]

**Comment by teor2345** [comment]
@yaahc I'm done with the implementation now, so there shouldn't be any more RFC changes.

(Or if there are, I'll open a new pull request.)

**Comment by mpguerra** [comment]
Do we need another review from @yaahc or are we good to merge this one now?

**Review by yaahc — APPROVED** [review]
:+1:
